### PR TITLE
[fix]: [OIP-833]: Extra validation on Datadog verification

### PR DESCRIPTION
**Pull request by AbhijithHarness**
## Describe your changes
Just adding extra validation on Datadog in Backend, Validating required fields are present in case of deployment verification

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/52284)

### CODE REVIEW

**Comment by AbhijithHarness** [comment]
trigger uts

**Review by dhruv-harness — COMMENTED** [review]
Can we add a UT for this?

**Comment by AbhijithHarness** [comment]
> 

asdiscussed with Sunil, we will add it later.

**Review by dhruv-harness — APPROVED** [review]

**Comment by AbhijithHarness** [comment]
trigger messagecheck

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java [hunk 1]
```diff
@@ -65,6 +66,14 @@ public void populateFromMetricDefinitions(
                                 .build();
 
     datadogMetricDefinitions.forEach(definition -> {
+      if (definition.getAnalysis() != null && definition.getAnalysis().getDeploymentVerification() != null
+          && definition.getAnalysis().getDeploymentVerification().getEnabled() != null
+          && definition.getAnalysis().getDeploymentVerification().getEnabled()) {
+        Preconditions.checkArgument(StringUtils.isNotEmpty(definition.getGroupingQuery())
+                && StringUtils.isNotEmpty(definition.getServiceInstanceIdentifierTag()),
+            "metric: %s, serviceInstanceIdentifierTag and groupingQuery are required for Deployment verification,"
```
Comment (n00bitax) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: we cant put this validation

Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: Grouping query is required for deployment verfications right?

**Comment by n00bitax** [comment]
groupingQuery is optional, and is not required always

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java (hunk 1 above)
Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: no

Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: groupingQuery is formed by the FE and it is not able to do that formulas (composed of multiple queries)

Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: when it is required?

Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: and not?

Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: After DATADOG FF is removed we will stop using the groupingQuery permanently

Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: So how can we verify that the Grouping query is added by the customer?

Comment (AbhijithHarness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: for FF off?

Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: We can keep serviceInstanceIdentifierTag validation

Comment (n00bitax) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/DatadogMetricCVConfig.java:74]: Customer never adds any grouping Query its formed by FE and sent in background

**Review by n00bitax — APPROVED** [review]

**Comment by n00bitax** [comment]
trigger uts

**Pull request merged**